src/apigateway/router.js

```diff
@@ -43,6 +44,9 @@ class Router {
                 response: this._errors
             });
         }
+        if (typeof this._onError === 'function') {
+            this._onError(error, this._errors);
+        }
```

Comment:
The function signature should be the same as all the other function contexts that we use. If you look at the `_beforeAll` and `_afterAll`, we pass:

```javascript
await this._beforeAll(request, response, endpoint.requirements);
await this._afterAll(request, response, endpoint.requirements);
```
Obviously, we don't need the requirements on this, but we should still be passing in request and response objects, followed by an additional error object.
 
We should be doing something like:

```suggestion
        if (typeof this._onError === 'function') {
            this._onError(request, response, error);
        }
```

You will need to mutate the response object to be able to get a response out to the API end user. 

Lastly, I don't think we should run both this and if statements above L33 `if (!this._errors.hasErrors)` it should be one or the other, we don't want mutate the error object with our standard response if the user is writing their own error handler, this whole handle method error needs to be refactored, something like:


```javascript
const request = new RequestClient(this._event);
const response = new ResponseClient();
if (typeof this._onError === 'function') {
    this._onError(request, response, error);
} else if (!this._errors.hasErrors) {
    this._errors.code = 500;
    this._errors.setError('server', 'internal server error');
}
if (!process.env.unittest) {
            this._logger.error({
                error_messsage: error.message,
                error_stack: error.stack instanceof String ? error.stack.split('\n') : error,
                event: this._event,
                request: request.request,
                response: this._errors
      });
}
```

Reply:
Thanks for the so detailed review. I've fixed all issues.